### PR TITLE
Lasso and PPCA

### PR DESCRIPTION
**Pull request by SvenCelin**
Lasso and PPCA functions transferred in Builtins

### CODE REVIEW

**Review by Baunsgaard — COMMENTED** [review]
Hi @SvenCelin,

Thanks for adding PPCA and Lasso to the builtin algorithms. :medal_sports: 
There is some descriptions and details missing, but i hope it is not to much work.

Please talk together with @gabrielaozegovic , you both are having the same test issues. I would suggest that you work together to fix one, then the others will be easier.

File: scripts/builtin/lasso.dml [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+# specific language governing permissions and limitations
+# under the License.
+#
+#-------------------------------------------------------------
+
+#uses the sparsa algorithm to perform lasso regression
+
+
+m_lasso = function(Matrix[Double] X, Matrix[Double] y) return(Matrix[Double] w)
+    {
```
[the remark this answers is not in the document]
Comment (SvenCelin) [reply scripts/builtin/lasso.dml]: Done.

File: scripts/builtin/lasso.dml [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+#-------------------------------------------------------------
+
+#uses the sparsa algorithm to perform lasso regression
+
```
[the remark this answers is not in the document]
Comment (SvenCelin) [reply scripts/builtin/lasso.dml:22]: Done.

File: scripts/builtin/lasso.dml [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+#-------------------------------------------------------------
+
+#uses the sparsa algorithm to perform lasso regression
```
[the remark this answers is not in the document]
Comment (SvenCelin) [reply scripts/builtin/lasso.dml]: It was a typo, I have changed it in the description now.

File: scripts/builtin/lasso.dml [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+
+m_lasso = function(Matrix[Double] X, Matrix[Double] y) return(Matrix[Double] w)
+    {
+    n = nrow(X)
+    m = ncol(X)
+
+
+    #params
```
[the remark this answers is not in the document]
Comment (SvenCelin) [reply scripts/builtin/lasso.dml]: Done.

File: scripts/builtin/lasso.dml [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    tol = 10^(-15)
+    M = 5
+    tau = 1
+    maxiter = 1000
+
+    #constants
```
[the remark this answers is not in the document]
Comment (SvenCelin) [reply scripts/builtin/lasso.dml]: They are constants, I don't think the tampering with them would do much good.

File: scripts/builtin/ppca.dml [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+# ---------------------------------------------------------------------------------------------
+# NAME   	TYPE   DEFAULT  MEANING
+# ---------------------------------------------------------------------------------------------
+# X  	 	String ---      location to read the matrix X input matrix
+# k      	Int    ---      indicates dimension of the new vector space constructed from eigen vectors
+# tolobj 	Int    0.00001  objective function tolerance value to stop ppca algorithm
+# tolrecerr	Int    0.02     reconstruction error tolerance value to stop the algorithm
+# iter   	Int    10       maximum number of iterations
+# fmt    	String 'text'   output format of results PPCA such as "text" or "csv"
+# hadoop jar SystemDS.jar -f PPCA.dml -nvargs X=/INPUT_DIR/X  C=/OUTPUT_DIR/C V=/OUTPUT_DIR/V k=2 tol=0.2 iter=100
```
[the remark this answers is not in the document]
Comment (SvenCelin) [reply scripts/builtin/ppca.dml]: Done.

File: scripts/builtin/ppca.dml [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+        #Frobenius norm of reconstruction error -> Euclidean norm
+        #Fn -> 1 x 1
+        Fn = sum(Xm*Xm);
+
+        #ss -> 1 x 1
+        ss = (Fn + ss2 - 2*ss3)/(n*m);
+
+       #calculating objective function relative change
+       ObjRelChng = abs(1 - ss/ssPrev);
+       #print("Objective Relative Change: " + ObjRelChng + ", Objective: " + ss);
```
[the remark this answers is not in the document]
Comment (SvenCelin) [reply scripts/builtin/ppca.dml]: Done.

File: scripts/builtin/ppca.dml [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+# OUTPUT PARAMETERS:
+# ---------------------------------------------------------------------------------------------
+# NAME   TYPE   DEFAULT  MEANING
+# ---------------------------------------------------------------------------------------------
+# C     	Matrix  ---     principal components
+# V      	Matrix  ---     eigenvalues / eigenvalues of principal components
+#
+
+
+m_pcaa = function(Matrix[Double] X, int iter = 10, double tolobj = 0.00001, double tolrecerr = 0.02, String fmt0 = "text") return(Matrix[Double] PC, Matrix[Double] V)
```
[the remark this answers is not in the document]
Comment (SvenCelin) [reply scripts/builtin/ppca.dml]: Done.

File: scripts/builtin/ppca.dml [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+# INPUT PARAMETERS:
+# ---------------------------------------------------------------------------------------------
+# NAME   	TYPE   DEFAULT  MEANING
+# ---------------------------------------------------------------------------------------------
+# X  	 	String ---      location to read the matrix X input matrix
+# k      	Int    ---      indicates dimension of the new vector space constructed from eigen vectors
+# tolobj 	Int    0.00001  objective function tolerance value to stop ppca algorithm
+# tolrecerr	Int    0.02     reconstruction error tolerance value to stop the algorithm
+# iter   	Int    10       maximum number of iterations
+# fmt    	String 'text'   output format of results PPCA such as "text" or "csv"
```
[the remark this answers is not in the document]
Comment (SvenCelin) [reply scripts/builtin/ppca.dml]: Done.

File: src/main/java/org/apache/sysds/common/Builtins.java [hunk 10]
```diff
@@ -152,6 +153,7 @@
 	PCA("pca", true),
 	PNMF("pnmf", true),
 	PPRED("ppred", false),
+	PPCA("ppca", true),
```
[the remark this answers is not in the document]
Comment (SvenCelin) [reply src/main/java/org/apache/sysds/common/Builtins.java:155]: Done.

File: src/test/java/org/apache/sysds/test/functions/builtin/BuiltinLassoTest.java [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    private void runLassoTest(){
+
+        loadTestConfiguration(getTestConfiguration(TEST_NAME));
+        String HOME = SCRIPT_DIR + TEST_DIR;
+        fullDMLScriptName = HOME + TEST_NAME + ".dml";
+        List<String> proArgs = new ArrayList<>();
+
+
+        proArgs.add("-explain");
+        proArgs.add("-stats");
```
[the remark this answers is not in the document]
Comment (SvenCelin) [reply src/test/java/org/apache/sysds/test/functions/builtin/BuiltinLassoTest.java]: Done.

File: src/test/scripts/functions/builtin/PPCA.dml [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+#-------------------------------------------------------------
+
+X = read($1)
+PC = ppca(X = X)
+write(PC, V)
+
```
[the remark this answers is not in the document]
Comment (SvenCelin) [reply src/test/scripts/functions/builtin/PPCA.dml:26]: Done.

File: src/test/scripts/functions/builtin/lasso.dml [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+#-------------------------------------------------------------
+
+X = read($1)
+y = read($2)
+w = lasso(X = X, y = y)
+write(w, $3)
```
[the remark this answers is not in the document]
Comment (SvenCelin) [reply src/test/scripts/functions/builtin/lasso.dml:25]: Done.

File: scripts/builtin/ppca.dml [hunk 14]
```diff
@@ -0,0 +1,154 @@
...
+    # sorting eigenvalues and eigenvectors in decreasing order
+    V_decr_idx = order(target=V,by=1,decreasing=TRUE,index.return=TRUE);
+    VF_decr = table(seq(1,nrow(V)),V_decr_idx);
+    V = VF_decr %*% V;
+    PC = PC %*% VF_decr;
+
+    # writing principal components
+    # write(PC, fileC, format=fmt0);
+    # writing eigen values/pc variance
+    # write(V, fileV, format=fmt0);
```
[the remark this answers is not in the document]
Comment (SvenCelin) [reply scripts/builtin/ppca.dml]: Done.

File: src/test/java/org/apache/sysds/test/functions/builtin/BuiltinLassoTest.java [hunk 15]
```diff
@@ -0,0 +1,55 @@
+package org.apache.sysds.test.functions.builtin;
```
[the remark this answers is not in the document]
Comment (SvenCelin) [reply src/test/java/org/apache/sysds/test/functions/builtin/BuiltinLassoTest.java:1]: Done.

File: src/test/java/org/apache/sysds/test/functions/builtin/BuiltinPPCATest.java [hunk 16]
```diff
@@ -0,0 +1,49 @@
+package org.apache.sysds.test.functions.builtin;
```
[the remark this answers is not in the document]
Comment (SvenCelin) [reply src/test/java/org/apache/sysds/test/functions/builtin/BuiltinPPCATest.java:20]: Done.

**Comment by mboehm7** [comment]
LGTM - thanks for the patch @SvenCelin. This completes the AMLS project.

During the merge, I however, made a number of modifications: (1) added test assertions on expected results, (2) fixed the formatting of tests (tabs over spaces for java code), (3) fixed the integration of ppca (the failure was due to a typo: pcaa instead of ppca), (4) modified the outputs of ppca to match the outputs of pca, (5) modified the parameters of ppca (e.g., passing of k), (6) added documentation for the parameters of lasso, and (7) fixed the formatting of builtin functions (2 space indentation).

**Pull request closed without merging**